### PR TITLE
feat(core): score a category by how much is wrong, not whether anything is

### PR DESCRIPTION
**Pull request by oekazuma**
## Why

A category score moved with **whether** something was wrong, not with **how much**.

Two symptoms, both measured. They share one cause.

**A category could not reach most of its own scale.** A key — a route or a source file — lost a fixed number of points per failing rule, and is only ever touched by rules of its own scope. So the lowest score a key could reach was fixed by that scope's rule inventory:

| category / scope | rules | lowest reachable |
| --- | --- | --- |
| architecture / component | 8 info | **92** |
| seo / component | 1 warning | **95** |
| performance / component | 4 info, 1 warning | **91** |
| performance / route | 3 info, 5 warning | 72 |
| security / component | 4 warning, 1 critical | 65 |
| correctness / component | 1 info, 10 warning, 3 critical | 4 |
| seo / route | 10 info, 14 warning, 2 critical | 0 |

Architecture is eight `info` rules and nothing else, so an architecture score was a nine-value scale presented as a hundred-value one. Three more scopes sat above 90 — this was not one miscalibrated category, it is what the model did wherever a scope's rules were few or cheap.

**The average erased magnitude.** One `info` finding moves a mean of N keys by `1/N`. At the scale a field measurement reported — 585 keys — one finding and 276 findings both displayed **99**.

Fixing either symptom alone fixes neither. Raising the deductions leaves architecture bounded; un-bounding architecture without making each affected key cost more leaves the mean pinned near 100.

This is the follow-up the previous scoring change recorded against itself: "Trading '100 is a lie' for '99 says nothing' is the right trade only if the second half gets fixed."

## What

A key now scores the share of what it was measured against that is intact:

```
keyScore = inventoryWeight === 0 ? 100 : clamp(100 − (100 × failedWeight) / inventoryWeight)
```

`inventoryWeight` is `max(observedInventory, failedWeight)`, where `observedInventory` sums `DEDUCTION[severity]` over the selected rules whose **`(category, scope)` pair** is among the pairs observed on that key. Keys are averaged in **deficit space**. Every category can now reach 0, and the 276-finding case reads **94** against **99** for one finding.

`seo` and `correctness` — the largest inventories, and the categories users read most — land within a point of their old values. The change concentrates where the scale was most compressed.

### Decisions that are load-bearing rather than stylistic

**The denominator is the scope's inventory, not the rules that applied.** The faithful "not applicable" semantics collapses here: measuring each architecture rule's `applies` condition shows an ordinary component has **one or two** applicable architecture rules, so a props-less component a few lines over the size threshold would score **0**. Flooring the denominator does not rescue it — at 300 keys with 10 affected, a per-key 0 yields a category mean of 96.7 and a per-key 80 yields 99.3, so the larger the floor the less of the fix survives.

**The pair, not the scope alone.** `computeScore` takes results, not a category, and three call sites pass multi-category sets — `routes[].score`, the console's route tree, and the Vite plugin. Summing over observed `(category, scope)` pairs defines those and reduces to the single-category rule exactly when the input is single-category.

**Unchanged on purpose:** `sitePenalty` stays absolute points (neither symptom applies to it — it is never divided by the key count), a `critical` still caps a category at 79, and a displayed 100 still means no finding among the checks that ran.

**No signature cascade.** `computeScore` derives the inventory itself from `selectRules(allRules, config)`; `ScoreOptions` gains an optional `rules` for tests. Nothing under `src/rules/` imports anything under `src/scoring/`, so the direction is free. Deliberately unlike the previous change that threaded a rule-id list through `AnalyzeResult`: there the CLI's `--category` narrowing was needed; here it would be wrong, since `--category seo` must not shrink `seo`'s own inventory.

## The arithmetic, which is where this could be wrong while looking right

Three forms were mandatory, each because the alternative loses a displayed point:

- **`100 − (100 × f) / i`, never `100 × (1 − f / i)`** — the second yields `19.999999999999996` for `f = 88, i = 110` and displays 19 for a true 20.
- **The mean in deficit space** — a mean of key scores yields `49.99999999999999` for two keys at `f = 3` and `f = 6` against `i = 9`, displaying 49 for a true 50.
- **`max(observedInventory, failedWeight)` plus a zero guard** — without them a result whose rule is absent from the inventory produces `NaN`, which reaches the JSON reporter as `null`.

The design was reviewed adversarially three times and rejected twice before approval. That process caught, among others, two table values computed with rounding — the operation the predecessor design exists to forbid — and a justification for deficit space that was **arithmetically false**: a clean project's mean is exactly 100 under either form, so the test that justification implied could not fail. The real case is an exactly-integral mean of unclean keys. One residual is stated rather than smoothed: a multi-key exactly-integral mean may display one point low, bounded at one point and verified by exhaustive and randomized sweeps against exact rational truth.

## Verification

`core` 1210, `cli` 805, `vite` 206 tests pass; typecheck clean in all three after rebuilding core; lint and format clean.

Every guard was confirmed load-bearing by mutation, not assumed: the evaluation order, the deficit-space mean, `max`, the zero guard, the `?? 0` on the inventory lookup, the pair in both directions, and the inventory's dependence on configuration.

That last set exists because the whole-branch review found the opposite. **The test written to hold the evaluation order held nothing** — mutating the shipped code to either wrong form left all 1206 tests green, because the deficit-space mean turns `100 − 19.999999999999996` into exactly 80, rounding away the very error the anchor was chosen to expose. It is now anchored on `f = 28, i = 50` (44 shipped, 43 mutated), with the original kept as a labelled non-load-bearing example. Two further guards were found untested the same way and now have tests.

## Migration

**Every score changes, most of them downward, and by more than a point.** A `--min-health` gate calibrated against the old numbers will start failing; recalibrate against the new scale. `routes[].score` in the JSON report changes meaning the same way. Stored baselines are unaffected — they key on findings, not scores.

## Out of scope, found while reviewing

`pnpm smoke`'s JSON check fails deterministically on `main` as well as here: the fixture report is 67,656 bytes, past the 64 KiB pipe buffer, and the analyze path calls `process.exit` after `console.log`. Reproduced against `main` at 6174836c. It needs its own issue and is untouched by this branch.

Design: `docs/superpowers/specs/2026-08-04-score-proportionality-design.md`. Plan: `docs/superpowers/plans/2026-08-04-score-proportionality.md`.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Health scores now use severity-weighted proportional scoring based on applicable checks.
  * Scores can reach 0 and better reflect the number and severity of findings.
  * Duplicate findings count once per route and rule, using the highest severity.
* **Documentation**
  * Updated English and Japanese health-report and JSON reporter guidance.
  * Added guidance to recalibrate `--min-health` thresholds after upgrading.
* **Tests**
  * Expanded coverage for scoring boundaries, disabled rules, severity overrides, and edge cases.

### CODE REVIEW

File: packages/core/src/scoring/score.ts [hunk 1]
```diff
@@ -37,38 +42,49 @@ export function computeScore(results: Result[], config: Config, options: ScoreOp
   const routeResults = results.filter((r) => r.route !== undefined);
   const projectResults = results.filter((r) => r.route === undefined);
 
-  // Seed every route at 100 so passing routes count toward the average.
-  const routeScores = new Map<string, number>();
-  for (const r of routeResults) if (!routeScores.has(r.route as string)) routeScores.set(r.route as string, 100);
+  const rules = options.rules ?? selectRules(allRules, config);
+  const inventory = buildInventory(config, rules);
+  const pairOf = ruleScopes(rules);
```
[the remark this answers is not in the document]
Comment (oekazuma) [reply packages/core/src/scoring/score.ts]: Fixed in d8220d7e. Reproduced first, and the inconsistency is sharper than "scores 80": the same input gets two different answers depending on whether another rule happens to share the pair.

A `performance/component` list of one `off` `info` rule and one enabled `warning` rule, with only the `off` rule failing — injected alongside the enabled rule the key scored **80**, injected alone in its pair it scored **0**. The design's stated mechanism for this case ("it observes no pair, so it contributes nothing to `observedInventory`") only held in the second.

`computeScore` now runs a supplied list through `selectRules` itself, so `buildInventory` and `ruleScopes` see the same filtered rules. The mixed-pair case scores 0, matching the isolated one, and a regression test pins it — reverting the change fails that test with 80.

One consequence worth recording, because it was not obvious: once both helpers take the same list, every pair reachable through `pairOf` also has an inventory entry, so `inventory.get(p) ?? 0` becomes unreachable. Verified — with the fix applied, mutating it to `inventory.get(p)!` still passed the whole suite. The `?? 0` is kept as a degrade-to-0-rather-than-NaN fallback against a future divergence, with a comment saying so, and the test that previously claimed to hold that guard now asserts the observable behaviour instead.

The design doc's description of the `rules` escape hatch is updated to say a supplied list is not taken as given.

File: .changeset/score-proportionality.md [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+same score.
+
+A key now scores the share of what it was measured against that is intact, weighted by severity. Every
+category can reach 0, and the score moves with the number of findings.
+
+**Every score changes, most of them downward, and by more than a point.** `seo` and `correctness` stay
+within a point of their old values; `architecture`, `security` and `performance` move further, because their
+scales were the most compressed. A `--min-health` gate calibrated against the old numbers will start
+failing — recalibrate it against the new scale. `routes[].score` in the JSON report changes meaning the same
+way. Stored baselines are unaffected, since they key on findings rather than scores.
```
[the remark this answers is not in the document]
Comment (oekazuma) [reply .changeset/score-proportionality.md]: Fixed in 5f21d53c, though not with the proposed wording.

The underlying point is right and the sentence was literally false: a clean report scored 100 before and scores 100 now, so "every score changes" is wrong.

I did not adopt "may change, often downward and sometimes by more than a point" because a migration note that hedges is less useful than one that is precise, and the precise statement is available. It now reads: any category carrying a finding changes, most of them downward and by more than a point; a clean 100 stays 100. That keeps the actionable part — recalibrate a `--min-health` gate — while removing the false universal.

File: docs/superpowers/plans/2026-08-04-score-proportionality.md [hunk 3]
```diff
@@ -0,0 +1,679 @@
...
+
+export function pairKey(category: Category, scope: Scope): PairKey {
+  return `${category}::${scope}`;
+}
+
+/** A rule's severity as configured, matching how `selectRules` reads `config.rules`. */
+function severityOf(rule: Rule, config: Config): Severity {
+  const setting = settingSeverity(config.rules[rule.id]);
+  return setting !== undefined && setting !== 'off' ? setting : rule.severity;
+}
```
[the remark this answers is not in the document]
Comment (oekazuma) [reply docs/superpowers/plans/2026-08-04-score-proportionality.md:163]: Fixed in 5f21d53c. The contradiction is real: the plan's `severityOf` restores an `off` rule's own severity, which yields an inventory of 16, while the test three steps earlier asserts 15. The implementer hit it during Task 1 and resolved it the right way.

The historical text is kept as written and a dated supersession note now sits beside the code block, recording that the shipped `severityOf` returns `undefined` for an `off` rule and `buildInventory` skips it, and pointing at `packages/core/src/scoring/inventory.ts` and its test.

File: packages/core/test/score.test.ts [hunk 4]
```diff
@@ -276,3 +289,191 @@ describe('computeScore — a displayed 100 means zero deduction', () => {
...
+    const results = [
+      fail('s/c1', '/a', 'critical'),
+      fail('s/c2', '/a', 'critical'),
+      ...Array.from({ length: 11 }, (_, i) => fail(`s/w${i}`, '/a', 'warning')),
+      ...Array.from({ length: 3 }, (_, i) => fail(`s/i${i}`, '/a', 'info'))
+    ];
+    expect(computeScore(results, defineConfig({}), { rules, applyCriticalCap: false }).score).toBe(20);
+  });
+
+  it('keeps an integral score integral — load-bearing: this anchor does discriminate', () => {
```
[the remark this answers is not in the document]
Comment (oekazuma) [reply packages/core/test/score.test.ts]: Fixed in 5f21d53c. `AGENTS.md` states it directly — "Test names state the behaviour, not the reasoning" — and these three did not.

They now read `scores 44 for failedWeight 28 against inventory 50`, `scores 50 for a two-key mean of failed weight 2 and 9 against inventory 11`, and `scores 0 and finite for a rule turned off in config but still carried in options.rules`. The reasoning that was in the names — why these particular anchors discriminate against a wrong evaluation order, and why the earlier anchor did not — moved into the test bodies, where it is still worth having for the next reader.
